bdk_electrum/src/lib.rs

```diff
@@ -0,0 +1,800 @@
+        P: ChainPosition,
+        K: Ord + Clone,
+        S: IntoIterator<Item = (u32, Script)>,
+    {
+        let mut request_spks = params
+            .keychain_spks
+            .into_iter()
+            .map(|(k, s)| (k, s.into_iter()))
+            .collect::<BTreeMap<K, _>>();
+        let mut scanned_spks = BTreeMap::<(K, u32), (Script, bool)>::new();
+
+        let update = loop {
+            let mut update = self.prepare_update(&to_position, local_chain)?;
+
+            if !request_spks.is_empty() {
+                if !scanned_spks.is_empty() {
+                    let mut scanned_spk_iter = scanned_spks
+                        .iter()
+                        .map(|(i, (spk, _))| (i.clone(), spk.clone()));
+                    match self.populate_with_spks::<P, K, _, _>(
+                        &to_position,
+                        &mut update,
+                        &mut scanned_spk_iter,
+                        params.stop_gap,
+                        params.batch_size,
+                    ) {
+                        Err(InternalError::Reorg) => continue,
+                        Err(InternalError::ElectrumError(e)) => return Err(e.into()),
+                        Ok(mut spks) => scanned_spks.append(&mut spks),
+                    };
+                }
+                for (keychain, keychain_spks) in &mut request_spks {
+                    match self.populate_with_spks::<P, K, u32, _>(
+                        &to_position,
+                        &mut update,
+                        keychain_spks,
+                        params.stop_gap,
+                        params.batch_size,
+                    ) {
+                        Err(InternalError::Reorg) => continue,
+                        Err(InternalError::ElectrumError(e)) => return Err(e.into()),
+                        Ok(spks) => scanned_spks.extend(
+                            spks.into_iter()
+                                .map(|(spk_i, spk)| ((keychain.clone(), spk_i), spk)),
+                        ),
+                    };
+                }
+            }
+
+            if !params.txids.is_empty() {
+                match self.populate_with_txids(
+                    &to_position,
+                    &mut update,
+                    params.txids.iter().cloned(),
+                ) {
+                    Err(InternalError::Reorg) => continue,
+                    Err(InternalError::ElectrumError(e)) => return Err(e.into()),
+                    Ok(_) => {}
+                }
+            }
+
+            if !params.txs.is_empty() {
+                match self.populate_with_txs(&to_position, &mut update, params.txs.iter()) {
+                    Err(InternalError::Reorg) => continue,
+                    Err(InternalError::ElectrumError(e)) => return Err(e.into()),
+                    Ok(_) => {}
+                }
+            }
+
+            if !params.outpoints.is_empty() {
+                match self.populate_with_outpoints(
```

Comment:
All of these functions are called only once and so inlining them would make it easier to understand what is happening.

Also I think that we can remove `populate_with_spks` too. This will allow for removing `InternalError` and all these repeated match statements completely. Consider that for each loop iteration you can create an iterator of all the "already scanned" spks and the new ones and just take batches from there.

This comment is only a suggestion. I care much more about the public API!

Reply:
I think we can refactor in a separate PR for this. Although, I'm not yet sold on in-lining everything...